msal/src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -216,6 +219,21 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
                 }
             }
         }
+
+        private void UseFallBackTenantIDForB2C(AuthenticationRequestParameters requestParams, IdToken idToken)
+        {
+            //The B2C scenario does not return the tenantID within the ID Token at the moment.
+            //As a fall back, the tenantID is set to the tenantID that is in the B2C authority originally passed in by the user.
+
+            if (requestParams.Authority.AuthorityType == Core.Instance.AuthorityType.B2C)
+            {
+                var tenantID = requestParams.Authority.GetTenantId();
+
+                if (string.IsNullOrEmpty(idToken.TenantId) && !string.IsNullOrEmpty(tenantID))
```

Comment:
>string.IsNullOrEmpty(idToken.TenantId) [](start = 20, length = 38)

also you can move this up to be the first check made before checking for autorityType. If tenantid exists in the idtoken, nothing more is needed. This also means the method should probably be renamed to e.g.: EnsureTenantIDIsSpecifiedForB2CAuthority